src/Orleans/AssemblyLoader/CachedTypeResolver.cs

```diff
@@ -2,46 +2,38 @@
 using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Reflection;
+using Orleans.Utilities;
 
 namespace Orleans.Runtime
 {
-    internal class CachedTypeResolver
+    internal class CachedTypeResolver : ITypeResolver
     {
-        private readonly ConcurrentDictionary<string, Type> cache;
-
-        static CachedTypeResolver()
-        {
-            Instance = new CachedTypeResolver();
-        }
-        public static CachedTypeResolver Instance { get; private set; }
-
-        protected CachedTypeResolver()
-        {
-            cache = new ConcurrentDictionary<string, Type>();
-        }
+        private readonly ConcurrentDictionary<string, Type> typeCache = new ConcurrentDictionary<string, Type>();
+        private readonly CachedReadConcurrentDictionary<string, Assembly> assemblyCache = new CachedReadConcurrentDictionary<string, Assembly>();
+      
+        public static ITypeResolver Instance { get; } = new CachedTypeResolver();
```

Comment:
can this eventually go as a singleton in the container instead of static property?

Reply:
Yes, `ITypeResolver` is usually used as a singleton. This is currently used in 2 places:
* `ApplicationConfiguration.ValidateConfiguration(logger)`
*  `FilterPredicateWrapperData.RehydrateStaticFuncion`

We could new up an instance each time in the 1st case and in the 2nd case we could have a static instance on the class.

I'm not so sure about the 2nd case, and I'm not a fan of how stream predicates are implemented.